internal/resolution/dependency_subgraph.go

```diff
@@ -0,0 +1,244 @@
+				gNodes[node] = gNode
+				// Populate parent's children and add to next set.
+				for _, edge := range parents {
+					nID := edge.From
+					pNode := gNodes[nID]
+					pNode.Children = append(pNode.Children, edge)
+					gNodes[nID] = pNode
+					if _, ok := seen[nID]; !ok {
+						seen[nID] = struct{}{}
+						next = append(next, nID)
+					}
+				}
+			}
+			toProcess = next
+			currDistance++
+		}
+
+		subGraphs = append(subGraphs, &DependencySubgraph{
+			Dependency: nodeID,
+			Nodes:      gNodes,
+		})
+	}
+
+	return subGraphs
+}
+
+// IsDevOnly checks if this DependencySubgraph solely contains dev (or test) dependencies.
+// If groups is nil, checks the dep.Type of the direct graph edges for the Dev Attr (for in-place).
+// Otherwise, uses the groups of the direct dependencies to determine if a non-dev path exists (for relax/override).
+func (ds *DependencySubgraph) IsDevOnly(groups map[manifest.RequirementKey][]string) bool {
+	if groups != nil {
+		// Check if any of the direct dependencies are not in the dev group.
+		return !slices.ContainsFunc(ds.Nodes[0].Children, func(e resolve.Edge) bool {
+			req := resolve.RequirementVersion{
+				VersionKey: ds.Nodes[e.To].Version,
+				Type:       e.Type.Clone(),
+			}
+			ecosystem, ok := util.OSVEcosystem[req.System]
+			if !ok {
+				return true
+			}
+
+			return !lockfile.Ecosystem(ecosystem).IsDevGroup(groups[manifest.MakeRequirementKey(req)])
+		})
+	}
+
+	// groups == nil
+	// Check if any of the direct dependencies do not have the Dev attr.
+	for _, e := range ds.Nodes[0].Children {
+		if e.Type.HasAttr(dep.Dev) {
+			continue
+		}
+		if e.To == ds.Dependency {
+			return false
+		}
+		// As a workaround for npm workspaces, check for the a Dev attr in the direct dependency's dependencies.
+		for _, e2 := range ds.Nodes[e.To].Children {
+			if !e2.Type.HasAttr(dep.Dev) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// ConstrainingSubgraph tries to construct a subgraph of the subgraph that includes only the edges that contribute to a vulnerability.
+// It identifies the dependencies which constrain the vulnerable package to use a vulnerable version.
+// This is used by the 'relax' remediation strategy to identify which direct dependencies need to be updated.
+//
+// e.g. for a subgraph with:
+//
+//	A -> C@<2.0
+//	B -> C@<3.0
+//	C resolves to C@1.9
+//
+// If the vuln affecting C is fixed in version 2.0, the constraining subgraph would only contain A,
+// since B would allow versions >=2.0 of C to be selected if not for A.
+//
+// This is a heuristic approach and may produce false positives.
+// If the constraining subgraph cannot be computed for some reason, returns the original DependencySubgraph.
+func (ds *DependencySubgraph) ConstrainingSubgraph(ctx context.Context, cl resolve.Client, vuln *models.Vulnerability) *DependencySubgraph {
+	// Just check if the direct requirement of the vulnerable package is constraining it.
+	// This still has some false positives.
```

Comment:
What's the impact on our results of getting this wrong?

Reply:
It would usually mean some direct dependencies that did not actually need to be relaxed end up being relaxed (or if they can't be relaxed further, the vuln deemed 'unfixable'). I've briefly noted this in the outer comment.

FWIW, this isn't new behaviour - this comment & logic is from the removed `DependencyChain` code. I also think this situation is unlikely to occur with npm.